bigquery/google/cloud/bigquery/_pandas_helpers.py

```diff
@@ -285,6 +294,44 @@ def dataframe_to_bq_schema(dataframe, bq_schema):
                 bq_schema_unused
             )
         )
+
+    # If schema detection was not successful for all columns, also try with
+    # pyarrow, if available.
+    if any(field.field_type is None for field in bq_schema_out):
+        if not pyarrow:
+            return None  # We cannot detect the schema in full.
+
+        arrow_table = dataframe_to_arrow(dataframe, bq_schema_out)
```

Comment:
Converting the entire dataframe to pyarrow tables is not ideal, of course - is there a shortcut?

Thought: We might want to issue an extra warning here, warning about a possible negative performance impact for large dataframes? On the plus side, it's at least a last resort in schema autodetection, and not a main code path.

Reply:
It'd be slightly better to make the conversion only on the columns whose types we could not determine. Does the array function work okay when passed a column from a dataframe? https://arrow.apache.org/docs/python/generated/pyarrow.array.html#pyarrow.array

Also, I wonder if we should amend our return type to include the converted arrays? We'll have to do the same conversion later on, when we serialize the Parquet file. It'd be nice to avoid making the conversion twice. -- If we do this, then maybe we should just always do the conversion to Arrow arrays, anyway?

Reply:
Limiting ourselves only to the "problematic" columns makes sense, will check how it works.

I will also see if we can avoid unnecessary double conversions, good point.

Reply:
The code now only converts the columns that need extra schema detection effort. Passing a dataframe column to `pyarrow.array()` works smoothly.